### PR TITLE
Script and readability enhancements

### PR DESCRIPTION
**Pull request by Daniel-Fan**
This is a PR to
- Address several readability issues.
  - https://github.ibm.com/ibmprivatecloud/roadmap/issues/59943
  - https://github.ibm.com/ibmprivatecloud/roadmap/issues/60056
- Enhance the health check on `.status.installedCSV` for each subscription during the installation.
   - https://github.ibm.com/ibmprivatecloud/roadmap/issues/59973

### CODE REVIEW

File: cp3pt0-deployment/common/migrate_singleton.sh [hunk 1]
```diff
@@ -54,7 +54,7 @@ function main() {
     
     if [[ $ENABLE_LICENSING -eq 1 ]]; then
 
-        is_exists=$("$OC" get deployments ibm-licensing-operator -n "$OPERATOR_NS")
+        is_exists=$("$OC" get deployments ibm-licensing-operator -n "$OPERATOR_NS" --ignore-not-found)
```
Comment (Daniel-Fan) [inline cp3pt0-deployment/common/migrate_singleton.sh:57]: Handle error
```
Error from server (NotFound): deployments.apps "ibm-licensing-operator" not found
```

File: cp3pt0-deployment/common/migrate_singleton.sh [hunk 2]
```diff
@@ -84,13 +84,31 @@ function main() {
...
+        warning "Failed to restore IBMLicensing instance"
+    else
+        success "IBMLicensing instance is restored"
+    fi
 
 }
 
 function backup_ibmlicensing() {
+    create_namespace "${LICENSING_NS}"
+
```
Comment (Daniel-Fan) [inline cp3pt0-deployment/common/migrate_singleton.sh:106]: Handle error
```
Error from server (NotFound): error when creating "STDIN": namespaces "ibm-licensing" not found
```

File: cp3pt0-deployment/common/migrate_singleton.sh [hunk 3]
```diff
@@ -84,13 +84,31 @@ function main() {
...
     # extracts the previously saved IBMLicensing CR from ConfigMap and creates the IBMLicensing CR
     "${OC}" get cm ibmlicensing-instance-bak -n ${LICENSING_NS} -o yaml --ignore-not-found | "${YQ}" .data | sed -e 's/.*ibmlicensing.yaml.*//' | 
-    sed -e 's/^  //g' | oc apply -f -
+    sed -e 's/^  //g' | "${OC}" apply -f -
+    
+    if [[ $? -ne 0 ]]; then
+        warning "Failed to restore IBMLicensing instance"
+    else
+        success "IBMLicensing instance is restored"
+    fi
```
Comment (Daniel-Fan) [inline cp3pt0-deployment/common/migrate_singleton.sh:100]: Handle error
```
error: no objects passed to apply
```

File: cp3pt0-deployment/common/utils.sh [hunk 4]
```diff
@@ -233,13 +233,13 @@ function wait_for_operator() {
...
+    local condition="${OC} get subscription.operators.coreos.com -l operators.coreos.com/${package_name}.${namespace}='' -n ${namespace} -o yaml -o jsonpath='{.items[*].status.installedCSV}'"
     local retries=30
     local sleep_time=10
     local total_time_mins=$(( sleep_time * retries / 60))
-    local wait_message="Waiting for operator ${package_name} CSV in namespace ${namespace} to be generated"
-    local success_message="Operator ${package_name} CSV in namespace ${namespace} is generated"
-    local error_message="Timeout after ${total_time_mins} minutes waiting for ${package_name} CSV in namespace ${namespace} to be generated"
+    local wait_message="Waiting for operator ${package_name} CSV in namespace ${namespace} to be bundled with Subscription"
+    local success_message="Operator ${package_name} CSV in namespace ${namespace} is bundled with Subscription"
+    local error_message="Timeout after ${total_time_mins} minutes waiting for ${package_name} CSV in namespace ${namespace} to be bundled with Subscription"
```
Comment (Daniel-Fan) [inline cp3pt0-deployment/common/utils.sh:242]: This is used to check `.status.installedCSV`
https://github.ibm.com/ibmprivatecloud/roadmap/issues/59973

File: cp3pt0-deployment/common/utils.sh [hunk 5]
```diff
@@ -902,7 +902,7 @@ function get_control_namespace() {
     local config_map_name="common-service-maps"
 
     # Get the ConfigMap data
-    config_map_data=$(${OC} get configmap "${config_map_name}" -n kube-public -o yaml | ${YQ} '.data[]')
+    config_map_data=$(${OC} get configmap "${config_map_name}" -n kube-public -o yaml --ignore-not-found | ${YQ} '.data[]')
```
Comment (Daniel-Fan) [inline cp3pt0-deployment/common/utils.sh:905]: Handle error
```
Error from server (NotFound): configmaps "common-service-maps" not found
[✗] Not found common-serivce-maps ConfigMap in kube-public namespace. It is a single shared Common Service instance upgrade
```

File: cp3pt0-deployment/common/utils.sh [hunk 6]
```diff
@@ -1229,6 +1229,12 @@ function accept_license() {
         info "Preview mode is on, skip patching license acceptance\n"
         return 0       
     fi
+
+    if [[ -z "$(${OC} get $kind $cr_name -n $namespace --ignore-not-found)" ]]; then
+        warning "Not found $kind $cr_name in $namespace, skipping updating license acceptance\n"
+        return 0
+    fi
+
```
Comment (Daniel-Fan) [inline cp3pt0-deployment/common/utils.sh:1237]: Handle error
```
# Accepting license for namespacescope common-service in namespace ibm-common-services...
Error from server (NotFound): namespacescopes.operator.ibm.com "common-service" not found
[✗] Failed to update license acceptance for namespacescope CR common-service
```

File: cp3pt0-deployment/setup_singleton.sh [hunk 7]
```diff
@@ -200,7 +200,7 @@ function is_migrate_licensing() {
 
     title "Check migrating LTSR ibm-licensing-operator"
     
-    local version=$("$OC" get ibmlicensing instance -o jsonpath='{.spec.version}')
+    local version=$("$OC" get ibmlicensings.operator.ibm.com instance -o jsonpath='{.spec.version}' --ignore-not-found)
```
Comment (Daniel-Fan) [inline cp3pt0-deployment/setup_singleton.sh:203]: Handle error
```
# Check migrating LTSR ibm-licensing-operator
Error from server (NotFound): ibmlicensings.operator.ibm.com "instance" not found
[✗] No version field in ibmlicensing CR, skipping
```

File: cp3pt0-deployment/setup_singleton.sh [hunk 8]
```diff
@@ -287,7 +287,7 @@ function install_cert_manager() {
         error "There is no cert-manager-webhook pod running\n"
     fi
 
-    local api_version=$("$OC" get deployments -n "$webhook_ns" cert-manager-webhook -o jsonpath='{.metadata.ownerReferences[*].apiVersion}')
+    local api_version=$("$OC" get deployments -n "$webhook_ns" cert-manager-webhook -o jsonpath='{.metadata.ownerReferences[*].apiVersion}' --ignore-not-found)
```
Comment (Daniel-Fan) [inline cp3pt0-deployment/setup_singleton.sh:290]: Handle error
```
# Check migrating and deactivating LTSR ibm-cert-manager-operator
Error from server (NotFound): configmaps "common-service-maps" not found
[✗] Not found common-serivce-maps ConfigMap in kube-public namespace. It is a single shared Common Service instance upgrade
```

File: cp3pt0-deployment/setup_tenant.sh [hunk 9]
```diff
@@ -558,7 +559,11 @@ function install_cs_operator() {
     fi
     
     # Checking master CommonService CR status
-    wait_for_cscr_status "$OPERATOR_NS" "common-service"
+    if [ $PREVIEW_MODE -eq 0 ]; then
+        wait_for_csv "$OPERATOR_NS" "ibm-odlm"
+        wait_for_operator "$OPERATOR_NS" "operand-deployment-lifecycle-manager"
+        wait_for_cscr_status "$OPERATOR_NS" "common-service"
+    fi
```
Comment (Daniel-Fan) [inline cp3pt0-deployment/setup_tenant.sh:566]: It is to check the health of ODLM, identify if `.status.installedCSV` is ready

File: cp3pt0-deployment/migrate_tenant.sh [hunk 10]
```diff
@@ -69,9 +69,6 @@ function main() {
     # Migrate singleton services
     if [[ $ENABLE_LICENSING -eq 1 ]]; then
         arguments+="--enable-licensing"
-        if [[ "$CONTROL_NS" != "$OPERATOR_NS" ]]; then
-            arguments+=" -licensingNs $CONTROL_NS"
-        fi
```
Comment (Daniel-Fan) [inline cp3pt0-deployment/migrate_tenant.sh:74]: Handle error: https://github.ibm.com/ibmprivatecloud/roadmap/issues/60056#issuecomment-61928346
It asks `setup_singleton.sh` to determine the namespace for Licensing Service.
```
[✘] An ibm-licensing-operator already installed in namespace: cs-control, expected namespace is: ibm-licensing
```

File: cp3pt0-deployment/setup_singleton.sh [hunk 11]
```diff
@@ -200,20 +200,24 @@ function is_migrate_licensing() {
...
-        return 0
-    fi
-
-    local ns=$("$OC" get deployments -A | grep ibm-licensing-operator | cut -d ' ' -f1)
-    if [ -z "$ns" ]; then
-        info "No LTSR ibm-licensing-operator to migrate, skipping"
+        if [[ "$CUSTOMIZED_LICENSING_NAMESPACE" -eq 1 ]] && [[ "$ns" != "$LICENSING_NAMESPACE" ]]; then
+            error "An ibm-licensing-operator already installed in namespace: $ns, please do not set parameter '-licensingNs $LICENSING_NAMESPACE"
+        fi
+        LICENSING_NAMESPACE="$ns"
```
Comment (Daniel-Fan) [inline cp3pt0-deployment/setup_singleton.sh:220]: Enhancement to automatically pick up correct Licensing Service Namespace IFF 1) Licensing operator deployment exists in that namespace AND 2) License CR has version >= v4.x.

**Comment by Daniel-Fan** [comment]
This is ready for reviewing

**Review by YCShen1010 — APPROVED** [review]
/lgtm

**Pull request merged**